src/binaryheap.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Datastructures: GAP package providing common datastructures.
+ * Licensed under the GPL 2 or later.
+ */
+
+#include "binaryheap.h"
+
+static UInt s_isLess_RNam = 0;  // FIXME: init these
+static UInt s_data_RNam = 0;      // FIXME: init these
+
+
+static void _BinaryHeap_BubbleUp_C(Obj data, Obj isLess, Int i, Obj elm)
+{
+    if ( LEN_PLIST( data ) < i ) {
+        GROW_PLIST( data, i );
+        SET_LEN_PLIST( data, i );
+        CHANGED_BAG( data );
+    }
+
+    while (i > 1) {
+        Obj parent = ELM_PLIST(data, i >> 1);
+        if (False == CALL_2ARGS(isLess, parent, elm))
```

Comment:
An important optimization: if isLess equals LtOper, use LT here.

Probably should move the check out of the loop, too.